src/test/java/hudson/remoting/ClassRemotingTest.java

```diff
@@ -109,49 +110,91 @@ public Object call() throws Exception {
-                        throw new Exception(t);
-                    }
-                }
-            });
+            java.util.concurrent.Future<Object> f1 = scheduleCallableLoad(c);
+
+            try {
+                f1.get();
+            } catch (ExecutionException ex) {
+                // Expected
+            }
+
+            // verify that classes that we tried to load aren't irrevocably damaged and it's still available
+            assertEquals(String.class, channel.call(c).getClass());
+        } finally {
+            RemoteClassLoader.TESTING_CLASS_LOAD = null;
+        }
+    }
+
+    @Bug(36991)
+    public void testInterruptionOfClassReferenceCreation() throws Exception {
+        DummyClassLoader parent = new DummyClassLoader(TestLinkage.B.class);
+        final DummyClassLoader child1 = new DummyClassLoader(parent, TestLinkage.A.class);
+        final DummyClassLoader child2 = new DummyClassLoader(child1, TestLinkage.class);
+        final Callable<Object, Exception> c = (Callable) child2.load(TestLinkage.class);
+        assertEquals(child2, c.getClass().getClassLoader());
+        RemoteClassLoader.TESTING_CLASS_REFERENCE_LOAD = new Interrupt3thInvocation();
 
-            // This is so that the first two class loads succeed but the third fails.
-            // A better test would use semaphores rather than timing (cf. the test before this one).
-            Thread.sleep(2500);
+        try {
+            Future<Object> f1 = scheduleCallableLoad(c);
 
-            f1.cancel(true);
             try {
-                Object o = f1.get();
-                assertEquals(String.class, o.getClass());
-                /* TODO we really want to fail here, but this method gets run 4×, and the last 3× it gets to this point:
-                fail(o.toString());
-                */
-            } catch (CancellationException x) {
-                // good
+                f1.get();
+            } catch (ExecutionException ex) {
+                // Expected
             }
 
             // verify that classes that we tried to load aren't irrevocably damaged and it's still available
-            assertNotNull(channel.call(c));
+            assertEquals(String.class, channel.call(c).getClass());
         } finally {
-            RemoteClassLoader.TESTING = false;
+            RemoteClassLoader.TESTING_CLASS_REFERENCE_LOAD = null;
+        }
+    }
+
+    private Future<Object> scheduleCallableLoad(final Callable<Object, Exception> c) {
+        ExecutorService svc = Executors.newSingleThreadExecutor();
+        return svc.submit(new java.util.concurrent.Callable<Object>() {
+            public Object call() throws Exception {
+                try {
+                    return channel.call(c);
+                } catch (Throwable t) {
+                    throw new Exception(t);
+                }
+            }
+        });
+    }
+
+    private static class Interrupt3thInvocation implements Runnable {
```

Comment:
I thought this was leetspeak + typo at first, so should be made `Third` or `3rd` for clarity.